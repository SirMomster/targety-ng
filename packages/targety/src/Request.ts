import cookie from "cookie";
import get from "lodash.get";
import toString from "lodash.tostring";
import * as querystring from "querystring";
import URLParse from "url-parse";
import { ContentType } from "./common/enums";
import { HttpMethod } from "./common/types";
import { Metadata } from "./Metadata";

export interface LambdaIdentity {
    accountId: string | null;
    sourceIp: string;
    userAgent: string | null;
    [key: string]: string | null | boolean;
}

export interface LambdaProxyEvent {
    body: string | null;
    headers: { [name: string]: string };
    multiValueHeaders: { [name: string]: string[] };
    httpMethod: HttpMethod;
    path: string;
    pathParameters: { [name: string]: string } | null;
    queryStringParameters: { [name: string]: string } | null;
    requestContext: {
        identity: LambdaIdentity;
        requestId: string;
        stage: string;
    };
    resource: string;
    [key: string]: any;
}

export interface Cookies {
    [key: string]: string[];
}

/**
 * Turns a lambda proxy event into a request object.
 *
 * @class Request
 */
export class Request {
    public isAuthenticated: boolean;
    private body: object | string;
    private token = new Map<string, string>();
    public headers: { [key: string]: string[] };
    private contentType: ContentType;
    private cookies: Cookies;
    private ip: string;
    private params: { [key: string]: string };
    private query: { [key: string]: string };
    private path: string;
    private resource: string;
    private xhr: boolean;
    private method: HttpMethod;
    private referer: URLParse;
    private userAgent: string;
    private identity: LambdaIdentity;
    private rawLambdaEvent: LambdaProxyEvent;
    private stage: string;
    private origin: string;
    private requestId: string;

    private _metadata: Metadata = new Metadata();

    /**
     * @param {LambdaProxyEvent} apiGateWayProxyEvent The event passed to the lambda function
     * with lambda-proxy type integration
     */
    constructor(apiGateWayProxyEvent: LambdaProxyEvent) {
        /** Parsed and normalized headers */
        this.headers = this.parseHeaders(apiGateWayProxyEvent);
        this.contentType = this.getHeader("content-type") as ContentType;

        /** Parses json or urlencoded string to JSON body or null */
        this.body = this.parseBody(apiGateWayProxyEvent);

        /** * Parsed cookies or empty object */
        this.cookies = this.parseCookies(this.headers.cookie);

        this.ip = get(apiGateWayProxyEvent, "requestContext.identity.sourceIp") || "";

        /**
         * This property is an object containing properties mapped to the named route “parameters”. For example,
         * if you have the route /user/:name, then the “name” property is available as req.params.name.
         * This object defaults to {}.
         */
        this.params = get(apiGateWayProxyEvent, "pathParameters") || {};

        /** * Passed query string parameters. Defaults to {}. */
        this.query = get(apiGateWayProxyEvent, "queryStringParameters") || {};

        /** * Contains the path part of the request URL. */
        this.path = get(apiGateWayProxyEvent, "path") || "";

        /** * Contains the resource path. */
        this.resource = get(apiGateWayProxyEvent, "resource") || "";

        /**
         * A Boolean property that is true if the request’s X-Requested-With header field is “XMLHttpRequest”,
         * indicating that the request was issued by a client library such as jQuery.
         */
        this.xhr = false;

        this.method = (toString(get(apiGateWayProxyEvent, "httpMethod")).toUpperCase() || "GET") as HttpMethod;

        /** Parsed referring including parsed query */
        this.referer = URLParse(toString(this.headers.referer), true);

        /** User agent passed from API Gateway */
        this.userAgent = get(apiGateWayProxyEvent, "requestContext.identity.userAgent") || "";

        /** AWS Lambda Identity (source: apiGateway, Cognito, or custom). Can be extended for internal api purpose */
        this.identity = get(apiGateWayProxyEvent, "requestContext.identity");

        /** Api gateway returns the stage in it's context */
        this.stage = get(apiGateWayProxyEvent, "requestContext.stage");

        /** the request's ID which is a unique identifier used by AWS for an event */
        this.requestId = get(apiGateWayProxyEvent, "requestContext.requestId");

        /** The origin domain coming from the request */
        this.origin = this.getHeader("origin");

        /** Raw API Gateway event */
        this.rawLambdaEvent = apiGateWayProxyEvent;
    }

    /**
     * Returns the field from the requestContext object from AWS API Gateway
     *
     * Returns undefined if nothing is found.
     * The Referrer and Referer fields are interchangeable.
     * @param {string} propertyPath
     * @returns {string|object}
     */
    public getContext(propertyPath: string): string | object {
        return get(this.rawLambdaEvent, `requestContext.${propertyPath}`);
    }

    /**
     * Returns query param value filtered for 'null', 'false', true', etc
     *
     * Returns undefined if nothing is found.
     * @param {string} param
     * @typeparam T Request a queryParam of type T. If not specified a type cast is needed for anything but string.
     */
    public getQueryParam<T extends string | number | boolean>(param: string): T {
        return this.cast(this.query[param]) as T;
    }

    /**
     * Returns query param value filtered for 'null', 'false', true', etc
     *
     * Returns undefined if nothing is found.
     * @param {string} param
     */
    public getPathParam(param: string): string | boolean | string[] {
        return this.valueFilter(this.params[param]);
    }
    public getQueryParams<T>(defaults?: Partial<T>): T {
        return { ...((defaults as object) || {}), ...(this.query as object) } as T;
    }
    public getPathParams(): { [key: string]: string } {
        return this.params;
    }
    public getBody<T extends object>(defaults?: Partial<T>): T {
        return { ...((defaults as object) || {}), ...(this.body as object) } as T;
    }
    public getCookies(): Cookies {
        return this.cookies;
    }
    public getResource(): string {
        return this.resource;
    }
    public getPath(): string {
        return this.path;
    }
    public getMethod(): HttpMethod {
        return this.method;
    }
    public getHeaders(): { [key: string]: string[] } {
        return this.headers;
    }
    public getIp(): string {
        return this.ip;
    }
    public getUserAgent(): string {
        return this.userAgent;
    }

    /**
     * Returns the list of forwarded IP addresses
     */
    public getForwardedIps(): string[] {
        return this.headers["x-forwarded-for"] || [];
    }

    public getTokens(): Map<string, string> {
        return this.token;
    }

    /**
     * Returns the cookie value, case-insensitive
     *
     * Returns undefined if nothing is found.
     * @param {string} name
     */
    public getCookie(name: string): string {
        const [cookie] = this.cookies[name.toLowerCase()] || [];
        return cookie;
    }

    /**
     * Returns the cookie for a specific environment if found
     *
     * @param name the cookie name
     * @param env the environment for which to find a cookie
     * @param strip strip the environment when true
     * @returns string
     */
    public getEnvCookie(name: string, env: string, strip = true): string {
        const cookies = this.cookies[name.toLowerCase()];
        if (!cookies) {
            return;
        }

        const cookie = cookies.find((v) => v.startsWith(env + "---"));
        if (cookie && strip) {
            return cookie.replace(`${env}---`, "");
        }

        return cookie;
    }

    /**
     * Returns the header value, case-insensitive
     *
     * Returns undefined if nothing is found.
     * @param {string} field
     */
    public getHeader(field: string): string {
        const header: string[] = this.headers[field.toLowerCase()];
        return header && header.join(",");
    }

    /**
     * Parses and normalizes the headers
     * @param lambdaEvent
     */
    private parseHeaders(lambdaEvent: LambdaProxyEvent): { [key: string]: string[] } {
        const lambdaHeaders = { ...get(lambdaEvent, "multiValueHeaders") };

        const headers: { [key: string]: string[] } = Object.keys(lambdaHeaders).reduce(
            (result: { [key: string]: string[] }, key) => {
                result[key.toLowerCase()] = lambdaHeaders[key];
                return result;
            },
            {},
        );

        // enforce 'referer' too because the internet can't decide
        if (headers.referrer) {
            headers.referer = headers.referrer;
            delete headers.referrer;
        }

        return headers;
    }

    /**
     * Parses the cookie headers and returns the cookies
     * this method will preserve multiple values for one key.
     *
     * @param cookieHeaders the cookie headers to be parsed
     */
    private parseCookies(cookieHeaders: string[] = []): Cookies {
        const parsedCookies = cookieHeaders.join(";").split(";");

        const result = new Map<string, string[]>();
        parsedCookies.forEach((v) => {
            const parseResult = cookie.parse(v);
            const [key] = Object.keys(parseResult);

            if (!key) return;
            const k = key.toLowerCase();

            if (!result.has(k)) {
                result.set(k, []);
            }

            result.get(k).push(parseResult[key]);
        });

        return Object.fromEntries(result);
    }

    /**
     * Parses body
     * @param lambdaEvent
     */
    private parseBody(lambdaEvent: LambdaProxyEvent): object | string {
        const bodyString = get(lambdaEvent, "body") || null;

        if (!bodyString) {
            return null;
        }
        if (typeof bodyString !== "string") {
            return bodyString;
        }

        if (this.contentType && this.contentType.match(ContentType.JSON)) {
            try {
                return JSON.parse(bodyString);
            } catch (e) {
                throw new Error("Request body contains invalid JSON");
            }
        } else if (this.contentType && this.contentType.match(ContentType.URL_ENCODED)) {
            return querystring.decode(bodyString);
        }

        return bodyString;
    }

    /** Casts a query or path parameter string to a JS primitive type if possible.  */
    private cast(val: string): string | number | boolean {
        if (!val) {
            return undefined;
        }
        if (typeof val === "boolean" || typeof val === "number") {
            return val;
        }
        if (val.toLowerCase() === "true") {
            return true;
        }
        if (val.toLowerCase() === "false") {
            return false;
        }
        if (val.toLowerCase() === "null") {
            return null;
        }
        if (val.toLowerCase() === "undefined") {
            return undefined;
        }
        if (!Number.isNaN(parseFloat(val))) {
            return parseFloat(val);
        }
        return val;
    }

    /**
     * Converts 'null' to null, 'false' to false, etc
     * @param (val) val
     */
    private valueFilter(val: string | boolean | string[]): string | boolean | string[] {
        if (typeof val !== "string") {
            return val;
        }

        const testVal = val.toLowerCase();

        if (testVal === "true") {
            return true;
        }

        if (testVal === "false") {
            return false;
        }

        if (testVal === "null") {
            return null;
        }

        return val;
    }

    public getReferer(): URLParse {
        return this.referer;
    }

    /**
     * Returns the host which equals the api gateway url or if coming from the frontend the frontend url
     */
    public getHost(): string {
        return this.getHeader("host");
    }

    /**
     * Returns the Stage for backend calls based on the stage value from ApiGateway API
     */
    public getStage(): string {
        return this.stage;
    }

    /**
     * Returns the host with the appropriate stage
     * @param path optional path that will be appended, make sure to start it with a "/"
     */
    public getFullRequestHost(path = ""): string {
        return `${this.getForwardedProtocol() || "https"}://${this.getHost()}/${this.getStage()}${path}`;
    }

    /**
     * Returns the host with no stage or path
     */
    public getFullRequestBaseHost(): string {
        return `${this.getForwardedProtocol() || "https"}://${this.getHost()}`;
    }

    /**
     * Returns the full origin url basically prefixing http:// for localhost if not present
     */
    public getOrigin(): string {
        return this.getHeader("Origin");
    }

    /**
     * Returns the protocol (fi. https) if set by cloudfront
     */
    public getForwardedProtocol(): string {
        return this.getHeader("X-Forwarded-Proto");
    }

    /**
     * Returns true when the request is coming from localhost
     */
    public isLocalhost(): boolean {
        return !!this.getOrigin()?.includes("localhost");
    }

    public getRequestId(): string {
        return this.requestId;
    }

    /**
     * Returns the metadata for the given request.
     */
    public get metadata(): Metadata {
        return this._metadata;
    }
}
