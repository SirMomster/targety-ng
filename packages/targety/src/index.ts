import "reflect-metadata";

export { LambdaEntryPoint } from "./EntryPoint";
export { Handler, Routes } from "./Handler";
export { Request, LambdaProxyEvent, LambdaIdentity } from "./Request";
export { Response } from "./Response";
export { ResponseBody, ResponseBodyInput } from "./ResponseBody";
export * as Middlewares from "./middlewares";
export { Middleware } from "./MiddlewareHandler";
export * as Types from "./common/types";
export * from "./common/decorators";
export * as Enums from "./common/enums";
export { Metadata } from "./Metadata";
