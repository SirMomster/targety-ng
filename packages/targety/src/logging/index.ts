export * from "./AuditLogger";
export { LOGGER } from "./Logger";
