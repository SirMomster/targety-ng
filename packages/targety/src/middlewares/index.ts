export { auditLog } from "./auditLog.middleware";
