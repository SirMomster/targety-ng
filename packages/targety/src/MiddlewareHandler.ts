import { ResponseBody } from ".";
import { Request } from "./Request";
import { Promises } from "./utils";
import { ApiError } from "./errors";
import { Context } from "./common/interfaces";
import { Route } from "./Route";

/**
 * Definition of an middleware
 */
export type Middleware = (request: Request, action: Route, context?: Context) => Promise<MiddlewareResponse>;
/**
 * The return type for middleware that define a follow up object
 *
 * Can be used when fi. you need to discard something after the request has completed or failed
 */
export type MiddlewareFollowUp = {
    onError: (request: Request, error: ApiError | Error) => Promise<void>;
    onSuccess: (request: Request, response: ResponseBody) => Promise<ResponseBody>;
};
/**
 * The the expected responses of functions that are used as middleware
 */
export type MiddlewareResponse = void | ResponseBody | MiddlewareFollowUp;

/**
 * Checks if the given object is a MiddlewareFollowUp object
 *
 * @param middlewareFollowUp an object that could be a MiddlewareFollowUp object
 */
export const isMiddlewareFollowUp = (middlewareFollowUp: MiddlewareFollowUp): boolean => {
    return !!(middlewareFollowUp?.onError || middlewareFollowUp?.onSuccess);
};

/**
 * Handles all middleware related logic, like resolving the middleware promise chain.
 *
 * @remark only use this as a class instance when you are sure that it can be properly reused
 * The risk is in the storing of the middleware response and middleware followup arrays.
 * Which need to be kept in state until they can be resolved (onError, onSuccess) when we keep this on a state level
 * we need to clear these arrays after each new incoming request. To avoid previous requests impacting the current ones
 * as state can be shared between multiple requests.
 */
export class MiddlewareHandler {
    private route: Route;
    private middleware: Middleware[];

    private middlewareResponses: MiddlewareResponse[] = [];
    private middlewareFollowUps: MiddlewareFollowUp[] = [];

    public constructor(route: Route, middleware: Middleware[]) {
        this.route = route;
        this.middleware = middleware;
    }

    /**
     * Handles an incoming request and runs the defined middleware against it.
     * Then checks if there are actions to be taken:
     *
     * - prematureErrorResponse: did an error occur in the middleware then we throw this error
     * - prematureResponse: did a middleware return a response then we return this response
     *
     * - middlewareFollowUp: are there actions to be taken on success or failure keep them in state
     *   until they can be resolved.
     *
     * @param request the incoming AWS API Gateway request
     * @param context the context of the execution environment
     */
    public async handle(request: Request, context: Context): Promise<ResponseBody | void> {
        /**
         * Run all the middleware and catch errors and responses so that they can be verified
         * Stops resolving the chain after the first error occurred
         */
        this.middlewareResponses = await Promises.resolvePromiseChain(
            this.middleware.map((middleware: Middleware) => async (chainResults: any) => {
                if (chainResults.find((v: any) => v instanceof Error)) return;

                try {
                    return await middleware(request, this.route, context);
                } catch (ex) {
                    return ex;
                }
            }),
        );

        /**
         * Store followUps so that they can be resolved when the request has succeeded or failed
         */
        this.middlewareFollowUps = this.middlewareResponses.filter(isMiddlewareFollowUp) as MiddlewareFollowUp[];

        // Throw the prematureError when there is one
        const prematureError = this.middlewareResponses.find((resp: any) => resp instanceof Error);
        if (prematureError) {
            throw prematureError;
        }

        // return the prematureResponse when there is one
        const prematureResponse = this.middlewareResponses.find((resp: any) => resp instanceof ResponseBody);
        if (prematureResponse && !isMiddlewareFollowUp(prematureResponse as any)) {
            return prematureResponse as ResponseBody;
        }
    }

    /**
     * Handles all success followups for a successful api request
     *
     * @param request the AWS API Gateway request
     * @param response the response to be send back to the user
     */
    public async handleSuccessFollowUps(request: Request, response: ResponseBody): Promise<ResponseBody> {
        const result: Array<ResponseBody> = await Promises.resolvePromiseChain(
            this.middlewareFollowUps.map((followUp: MiddlewareFollowUp) => async () =>
                await followUp.onSuccess(request, response),
            ),
        );

        return result.pop();
    }

    /**
     * Handles all failure followups for a unsuccessful api request
     *
     * @param request the AWS API Gateway request
     * @param error the error thrown by the executed code
     */
    public async handleFailureFollowUps(request: Request, error: ApiError | Error): Promise<void> {
        return await Promises.resolvePromiseChain(
            this.middlewareFollowUps.map((followUp: MiddlewareFollowUp) => async () =>
                await followUp.onError(request, error),
            ),
        );
    }
}
