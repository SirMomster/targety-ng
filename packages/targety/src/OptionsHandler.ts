import { Request } from "./Request";
import { ResponseBody } from "./ResponseBody";
import { ForbiddenError } from "./errors";
import { LOGGER } from "./logging";
import { Route } from "./Route";

/**
 * Creates a proper regex to match with the origin of the incoming request
 *
 * @param entry the domain that you would like to whitelist
 */
const buildCorsRegexps = (entry: string): RegExp => new RegExp(`https?://[\\w\\.-]*.?${entry}$`);

/**
 * Handles logic for creating proper option call responses
 */
export class OptionsHandler {
    private domainRegexps: RegExp[];
    /**
     * Creates a new instance of the OptionsHandler
     *
     * @param allowedDomains a list of urls that is to be allowed for cors requests
     */
    public constructor(
        private allowedDomains: string[] = [],
        private allowedHeaders: string[] = [],
        private exposedHeaders: string[] = [],
        private allowCredentials: boolean = false,
    ) {
        this.domainRegexps = allowedDomains.map(buildCorsRegexps);
    }

    /**
     * Checks if the origin is a valid hostname that has been whitelisted
     *
     * @param origin the origin url
     * @throws [[ForbiddenError]] when the incoming host doesn't match anything in the whitelist
     */
    protected validateHostIncomingOrigin(origin: string): string {
        if (this.domainRegexps.length === 0) {
            return origin;
        }

        LOGGER.debug("Validating if origin is whitelisted: " + origin);
        if (!this.domainRegexps.some((v) => v.test(origin))) {
            LOGGER.error("Origin isn't allowed: " + origin);
            throw new ForbiddenError("origin isn't allowed");
        }

        return origin;
    }

    /**
     * Creates a proper response for an incoming options call.
     *
     * @param request the request from API Gateway parsed as a Request object
     * @param actions the actions that match the path for which the options call has been done.
     */
    public optionsResponse(request: Request, actions: Route[]): ResponseBody {
        const origin = this.validateHostIncomingOrigin(request.getOrigin());

        return {
            statusCode: 200,
            body: null,
            multiValueHeaders: {
                "Access-Control-Expose-Headers": [`${this.exposedHeaders.join(",")}`],
                "Access-Control-Allow-Methods": [
                    `${["OPTIONS", ...actions.map((v) => v.method.toUpperCase())].join(",")}`,
                ],
                "Access-Control-Allow-Headers": [`${this.allowedHeaders.join(",")}`],
                "Access-Control-Allow-Origin": [`${origin}`],
                "Access-Control-Allow-Credentials": [`${this.allowCredentials}`],
            },
        };
    }
}
