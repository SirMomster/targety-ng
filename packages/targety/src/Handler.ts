import { LOGGER } from "./logging";
import { CLASS_METADATA, ROUTES_METADATA } from "./common/Constants";
import { HttpMethod } from "./common/types";
import { Request } from "./Request";
import { Response } from "./Response";
import { ResponseBody } from "./ResponseBody";
import { OptionsHandler } from "./OptionsHandler";
import { CustomerErrorHandlerMetaData, CustomErrorHandlerType } from "./common/decorators/CustomerErrorHandler";
import { Middleware, MiddlewareHandler } from "./MiddlewareHandler";
import { Context } from "./common/interfaces";
import { InternalServerError, NotFoundError } from "./errors";
import { Strings } from "./utils";
import { Route, RouteConfig } from "./Route";
import { HandlerMetaData } from "./HandlerMetaData";
import { CorsMetaData } from "./common/decorators";

export interface Routes {
    [key: string]: Route;
}

interface HandlerInterface {
    [key: string]: (request: Request) => Promise<ResponseBody>;
}

export abstract class Handler implements HandlerInterface {
    [key: string]: any;
    /**
     * Middleware is executed before the handler code. There are two types of middleware:
     *
     * 1) middleware that modifies the request object (express.js like functionality)
     * 2) middleware that returns a [[ResponseBody]] and interrupts the flow
     * 3) middleware that returns an object containing an onError and onSuccess
     *    function which can be resolved after the execution of the handler code.
     *
     * @protected
     * @abstract
     * @type {Middleware[]}
     * @memberof Handler
     */
    protected abstract middleware: Middleware[] = [];
    private handlerMetaData: HandlerMetaData = this.getClassMetaData();

    /**
     * Parses the action from the request, executes the middleware
     * and invokes the correct action on the handler implementation.
     */
    public async handle(request: Request, context?: Context): Promise<ResponseBody> {
        let customErrorHandler: CustomErrorHandlerType<Handler> | undefined = undefined;
        let middleWareHandler: MiddlewareHandler;
        let route: Route;

        try {
            if (request.getMethod() === "OPTIONS") {
                return await this.handleOptions(request);
            }

            route = this.getRouteConfig(request);

            middleWareHandler = new MiddlewareHandler(route, this.middleware);
            customErrorHandler = route.getMetaData<CustomerErrorHandlerMetaData>().customErrorHandler;

            // return the response from the middleware when a response is given
            const prematureResponse = await middleWareHandler.handle(request, context);
            if (prematureResponse) {
                return prematureResponse;
            }

            const result = await this[route.name](request);
            return (await middleWareHandler.handleSuccessFollowUps(request, result)) || result;
        } catch (e) {
            // Allows the route itself to handle errors in their own way when required
            await middleWareHandler?.handleFailureFollowUps(request, e);
            const customErrorResponse = customErrorHandler && (await customErrorHandler(request, this, e));

            const finalError = customErrorResponse || Response.fromError(request, e);

            // only log stacks for internal server errors as an error
            finalError.statusCode === InternalServerError.STATUS_CODE
                ? LOGGER.error("An internal server error occurred", e)
                : LOGGER.warn("An known error occurred", e);

            return finalError;
        }
    }

    /**
     * Handles incoming option callbacks for the Handler.handle method
     *
     * @param request the incoming request
     */
    protected async handleOptions(request: Request): Promise<ResponseBody> {
        const actions = this.getOptionsConfig(request);
        const corsMetaData = this.handlerMetaData.getMetaData<CorsMetaData>();

        const config = {
            allowCredentials: corsMetaData.AllowCredentials,
            allowedHeaders: [...corsMetaData.AllowHeaders],
            exposedHeaders: [...corsMetaData.ExposedHeaders],
        };

        actions.forEach((rc) => {
            const meta = rc.getMetaData<CorsMetaData>();
            config.allowedHeaders.push(...(meta?.AllowHeaders || []));
            config.exposedHeaders.push(...(meta?.ExposedHeaders || []));
            if (!config.allowCredentials) config.allowCredentials = meta.AllowCredentials;
        });

        const optionsHandler = new OptionsHandler(
            [],
            config.allowedHeaders,
            config.exposedHeaders,
            config.allowCredentials,
        );

        return optionsHandler.optionsResponse(request, actions);
    }

    /**
     * Parses the request so that it can be handled by our OptionsHandler
     *
     * @param request the incoming request
     */
    protected getOptionsConfig(request: Request): Route[] {
        const path: string = request.getResource();

        if (!path) {
            throw new NotFoundError("Route undefined");
        }

        const routesMetadata = Reflect.getMetadata(ROUTES_METADATA, this) as Routes;
        const routeNames: string[] = Object.keys(routesMetadata).filter((a) =>
            Strings.equalsTrimmedCaseInsensitive(path, routesMetadata[a].path),
        );

        if (!routeNames || routeNames.length === 0) {
            throw new NotFoundError("Route undefined");
        }

        return routeNames.map((routeName: string) => {
            const requestedAction: RouteConfig = routesMetadata[routeName];

            return new Route({
                name: routeName,
                path: requestedAction.path,
                method: requestedAction.method,
                metaData: requestedAction,
            });
        });
    }

    /**
     * Parses [[HandlerAction]] from request
     * @throws {Error} Unknown Route error
     */
    protected getRouteConfig(request: Request): Route {
        const path: string = request.getResource();
        const method: HttpMethod = request.getMethod();

        if (!path || !method) {
            throw new NotFoundError("Route undefined");
        }

        const routesMetadata = Reflect.getMetadata(ROUTES_METADATA, this) as Routes;
        const routeName: string = Object.keys(routesMetadata).find(
            (a) =>
                Strings.equalsTrimmedCaseInsensitive(path, routesMetadata[a].path) &&
                Strings.equalsTrimmedCaseInsensitive(method, routesMetadata[a].method),
        );
        if (!routeName) {
            throw new NotFoundError("Route undefined");
        }

        const requestedAction: RouteConfig = routesMetadata[routeName];
        const route = new Route({
            name: routeName,
            path: requestedAction.path,
            method: requestedAction.method,
            metaData: requestedAction,
        });

        return route;
    }

    protected getClassMetaData(): HandlerMetaData {
        const classMetaData = Reflect.getMetadata(CLASS_METADATA, this.constructor);
        return new HandlerMetaData({ metaData: classMetaData });
    }
}
