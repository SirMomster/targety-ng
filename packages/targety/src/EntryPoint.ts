import { LOGGER } from "./logging";
import { Objects } from "./utils";
import { Context } from "./common/interfaces";
import { Handler } from "./Handler";
import { LambdaProxyEvent, Request } from "./Request";
import { Response } from "./Response";
import { ResponseBody } from "./ResponseBody";

/**
 * Represents the LambdaEntryPoint abstract class
 */

export abstract class LambdaEntryPoint {
    public handler: Handler;
    public initPromise: Promise<Handler>;

    public constructor() {
        // pre-loading the handler, even when no request was made yet
        this.initPromise = this.initializeHandler();
    }

    /**
     * HeartBeatResponse can be used to keep the Lambda warm using Cloudwatch Events
     */
    private static heartBeatResponse(request: Request): ResponseBody {
        LOGGER.debug("Heartbeat function");
        // Disable cors because
        return Response.noContent(request, { cors: false, defaultHeaders: false }).send();
    }

    /**
     * Entrypoint for the Api Gateway Proxy event.
     */
    public async handle(event: LambdaProxyEvent, context: Context): Promise<ResponseBody> {
        let request: Request;
        LOGGER.debug(event, context, "Lambda Event");

        // Makes sure that the response is sent straight away when callback is invoked
        // instead of waiting for the nodejs event loop to be empty. This is best practice
        // for nodeJS lambdas on AWS
        // See https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html
        if (context) {
            context.callbackWaitsForEmptyEventLoop = false;
        }

        try {
            request = new Request(event);

            if (Objects.isEmpty(event)) {
                return LambdaEntryPoint.heartBeatResponse(request);
            }
            return await this.handleRequest(request, context);
        } catch (e) {
            LOGGER.error(e, "Error handling request");
            return Response.fromError(request, e);
        }
    }

    /**
     * initializeHandler is used for example to open a DB connection or set up Dependency Injection
     * @returns handler Handler exposes api endpoint methods
     */
    protected abstract initializeHandler(): Promise<Handler>;

    private async handleRequest(request: Request, context: Context): Promise<ResponseBody> {
        if (!this.handler) {
            this.handler = await this.initPromise;
        }
        return this.handler.handle(request, context);
    }
}
