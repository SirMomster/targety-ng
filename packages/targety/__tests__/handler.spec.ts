/* eslint-disable @typescript-eslint/ban-ts-ignore */
import { ApiGateWayProxyEvent } from "@dev/test-helper";
import "reflect-metadata";
import { CORS, DefaultCORS, Get, LambdaProxyEvent, Request, ResponseBody } from "../src";
import { Handler } from "../src/Handler";
import { Middleware } from "../src/MiddlewareHandler";

const getProfileStub = jest.fn();

@DefaultCORS({ ExposedHeaders: [], AllowCredentials: true, AllowHeaders: ["x-api-key"] })
class HandlerImplementation extends Handler {
    protected middleware: Middleware[] = [];

    @Get("/test-endpont/{id}")
    @CORS({ ExposedHeaders: ["x-test"] })
    public async getProfile(req: Request) {
        return getProfileStub(req);
    }
}

let handlerImplementation: any;
let event: LambdaProxyEvent;
let request: any;

describe("Handler", () => {
    beforeEach(() => {
        handlerImplementation = new HandlerImplementation();
        event = (ApiGateWayProxyEvent.get() as unknown) as LambdaProxyEvent;
        event.httpMethod = "GET";
        event.resource = "/test-endpont/{id}";
        request = new Request(event);
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it("is an abstract class that needs extending", () => {
        expect(handlerImplementation).toBeInstanceOf(Handler);
    });

    describe("setValidDomains", () => {
        const originalValue = process.env.NO_ALLOWED_DOMAINS;
        beforeAll(() => {
            process.env.NO_ALLOWED_DOMAINS = "false";
        });

        afterAll(() => {
            process.env.NO_ALLOWED_DOMAINS = originalValue;
        });
    });

    describe("#handle", () => {
        it("returns 500 response when Routepath is unknown", async () => {
            event.httpMethod = "PATCH";
            request = new Request(event);
            const response = await handlerImplementation.handle(request);
            expect(response.statusCode).toEqual(404);
        });
        it("invokes correct path", async () => {
            const response = await handlerImplementation.handle(request);
            expect(getProfileStub).toHaveBeenCalledTimes(1);
            expect(getProfileStub).toHaveBeenCalledWith(request);
        });
        it("invokes options", async () => {
            event = (ApiGateWayProxyEvent.get() as unknown) as LambdaProxyEvent;
            event.httpMethod = "OPTIONS";
            event.resource = "/test-endpont/{id}";
            request = new Request(event);
            const response = await handlerImplementation.handle(request);
            expect(response).toMatchSnapshot();
        });
        it("applies middleware and continues to correct path if middleware does not contain response", async () => {
            const middlewareMock = jest.fn().mockResolvedValue("OK");
            handlerImplementation.middleware.push(middlewareMock);
            const response = await handlerImplementation.handle(request);
            expect(middlewareMock).toHaveBeenCalledTimes(1);
            expect(middlewareMock).toHaveBeenCalledWith(
                request,
                handlerImplementation.getRouteConfig(request),
                undefined,
            );
            expect(getProfileStub).toHaveBeenCalledTimes(1);
            expect(getProfileStub).toHaveBeenCalledWith(request);
        });
        it("applies middleware and returns middleware ApiResponse", async () => {
            const middlewareMock = jest.fn().mockResolvedValue(new ResponseBody({ statusCode: 406 }));
            handlerImplementation.middleware.push(middlewareMock);
            const response = await handlerImplementation.handle(request);
            expect(middlewareMock).toHaveBeenCalledTimes(1);
            expect(middlewareMock).toHaveBeenCalledWith(
                request,
                handlerImplementation.getRouteConfig(request),
                undefined,
            );
            expect(getProfileStub).not.toHaveBeenCalled();
            expect(response.statusCode).toEqual(406);
        });
    });

    describe("#getRouteConfig", () => {
        it("parses a routeConfig from a request", () => {
            const route = handlerImplementation.getRouteConfig(request);

            expect(route.method).toEqual("GET");
            expect(route.name).toEqual("getProfile");
            expect(route.path).toEqual("/test-endpont/{id}");
        });
        it("returns Undefined Route error when METHOD + PATH is not in routeConfig", () => {
            event.httpMethod = "PATCH";
            request = new Request(event);
            expect(() => handlerImplementation.getRouteConfig(request)).toThrow("Route undefined");
        });
        it("returns Undefined Route error when METHOD is missing", () => {
            delete request.method;
            expect(() => handlerImplementation.getRouteConfig(request)).toThrow("Route undefined");
        });
        it("returns Undefined Route error when PATH is missing", () => {
            event.httpMethod = "PATCH";
            request = new Request(event);
            expect(() => handlerImplementation.getRouteConfig(request)).toThrow("Route undefined");
        });
    });

    describe("#getOptionsConfig", () => {
        it("should get the required config for an option call", () => {
            const routes = handlerImplementation.getOptionsConfig(request);

            expect(routes.length).toEqual(1);

            const route = routes[0];

            expect(route.method).toEqual("GET");
            expect(route.name).toEqual("getProfile");
            expect(route.path).toEqual("/test-endpont/{id}");
        });
        it("returns Undefined Route error when PATH is missing", () => {
            request = new Request(event);
            delete request.resource;
            expect(() => handlerImplementation.getRouteConfig(request)).toThrow("Route undefined");
        });
    });
});
