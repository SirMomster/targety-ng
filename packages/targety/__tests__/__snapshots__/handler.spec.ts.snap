// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handler #handle invokes options 1`] = `
Object {
  "body": null,
  "multiValueHeaders": Object {
    "Access-Control-Allow-Credentials": Array [
      "true",
    ],
    "Access-Control-Allow-Headers": Array [
      "x-api-key",
    ],
    "Access-Control-Allow-Methods": Array [
      "OPTIONS,GET",
    ],
    "Access-Control-Allow-Origin": Array [
      "http://localhost:3000",
    ],
    "Access-Control-Expose-Headers": Array [
      "x-test",
    ],
  },
  "statusCode": 200,
}
`;
